### ISSUE TITLE
Feature: Non-Linear Stretch (to Full Screen)

### ISSUE THREAD

**Issue by sevdestruct** [jasarien/Provenance#639]
> stretch game to fullscreen 16:9. ( with aspect ratio stretched to fill all screen)
(original request extracted from comment for [per-core CRT toggles] request (https://github.com/jasarien/Provenance/issues/634) made by @larssteenhoff)

**Full Request** (with non-linear addition):
Setting option stretch to full screen (as in fitting to full width of a display). Highly recommend doing this as non-linear stretch so it maintains the integrity of the art on screen, centrally, maximizing the screen equity without making all the characters look fat/wide as linear stretch would do uniformly.

**Comment by larssteenhoff** [comment]
Ideally both linear and non linear strech are supported.

For example I play f-zero many times full fullscreen stretch and yes there is distortion but for this game i dont care because its a car.

same for many other games. a custom aspect ration pixel value can be used too. this would accomodate both options.

**Comment by larssteenhoff** [comment]
video reference:

https://youtu.be/t7_NSTJ1ZX0

**Comment by sevdestruct** [comment]
Best Example I know of of Non-Linear Stretch is from Plex Media Player. Here are some snaps to illustrate what I mean. Fatness effect of linear stretch is worse with human characters, of course, but Teenage Mutant Ninja Turtles will be good enough to illustrate the point.

**Scaled:**
![c1a - scale](https://user-images.githubusercontent.com/3118097/34471938-7c2301b8-ef0a-11e7-97e3-80c436a74a36.png)

**Stretch (Linear):**
![c1a - stretch - linear](https://user-images.githubusercontent.com/3118097/34471940-85f217c4-ef0a-11e7-9937-c6db1a341c34.png)

**_Stretch (Non-linear):_**
![c1a - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471942-88814ef6-ef0a-11e7-8a37-9d9b609b0db4.png)

----

**Scaled:**
![c2a - scale](https://user-images.githubusercontent.com/3118097/34471949-bdab35ec-ef0a-11e7-8170-65130fdf290e.png)

**Stretch (Linear):**
![c2a - stretch - linear](https://user-images.githubusercontent.com/3118097/34471952-c01555e2-ef0a-11e7-8148-0db928647662.png)

**_Stretch (Non-linear):_**
![c2a - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471954-c6c2c4c4-ef0a-11e7-96eb-64a90948ef0a.png)

----

**Scaled:**
![a1 - scale](https://user-images.githubusercontent.com/3118097/34471957-d3a6a782-ef0a-11e7-9e40-d5bf516269ca.png)

**Stretch (Linear):**
![a2 - stretch - linear](https://user-images.githubusercontent.com/3118097/34471959-d56333d8-ef0a-11e7-8446-79fbd9fbb508.png)

**_Stretch (Non-linear):_**
![a3 - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471961-d8575d4e-ef0a-11e7-9d9b-454561d7b9a6.png)

----

**Stretch (Linear):**
![b2 - stretch - linear](https://user-images.githubusercontent.com/3118097/34471963-f303dbe0-ef0a-11e7-9873-7d361739022f.png)

**_Stretch (Non-linear):_**
![b1 - streth - non-linear](https://user-images.githubusercontent.com/3118097/34471965-f601a818-ef0a-11e7-9796-eff45aa5a3a7.png)

**Comment by sevdestruct** [comment]
Mocked up a sample of Chrono Trigger using a warp to simulate a non-linear stretch (8:7 -> 16:9):

Scaled:
![chrono-scaled](https://user-images.githubusercontent.com/3118097/34472136-910a5c16-ef0f-11e7-8006-df5fd6d268f2.jpg)

Stretched: Linear):
![chrono-stretched](https://user-images.githubusercontent.com/3118097/34472137-9938bbd0-ef0f-11e7-9e90-3fcd2dfbcc35.jpg)

Stretched: **_Non-linear_**:
![chrono-stretched-nonlinear](https://user-images.githubusercontent.com/3118097/34472138-9e037dda-ef0f-11e7-94ab-8470a711c696.jpg)

**Comment by sevdestruct** [comment]
Best Example I know of of Non-Linear Stretch is from Plex Media Player. Here are some snaps to illustrate what I mean. Fatness effect of linear stretch is worse with human characters, of course, but Teenage Mutant Ninja Turtles will be good enough to illustrate the point.

The  linear stretch tends to be more offensive when starting from something more narrow (like the SNES ratio of 8:7 versus the 4:3 ratio of the show below)…

**Scaled:**
![c1a - scale](https://user-images.githubusercontent.com/3118097/34471938-7c2301b8-ef0a-11e7-97e3-80c436a74a36.png)

**Stretch (Linear):**
![c1a - stretch - linear](https://user-images.githubusercontent.com/3118097/34471940-85f217c4-ef0a-11e7-9937-c6db1a341c34.png)

**_Stretch (Non-linear):_**
![c1a - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471942-88814ef6-ef0a-11e7-8a37-9d9b609b0db4.png)

----

**Scaled:**
![c2a - scale](https://user-images.githubusercontent.com/3118097/34471949-bdab35ec-ef0a-11e7-8170-65130fdf290e.png)

**Stretch (Linear):**
![c2a - stretch - linear](https://user-images.githubusercontent.com/3118097/34471952-c01555e2-ef0a-11e7-8148-0db928647662.png)

**_Stretch (Non-linear):_**
![c2a - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471954-c6c2c4c4-ef0a-11e7-96eb-64a90948ef0a.png)

----

**Scaled:**
![a1 - scale](https://user-images.githubusercontent.com/3118097/34471957-d3a6a782-ef0a-11e7-9e40-d5bf516269ca.png)

**Stretch (Linear):**
![a2 - stretch - linear](https://user-images.githubusercontent.com/3118097/34471959-d56333d8-ef0a-11e7-8446-79fbd9fbb508.png)

**_Stretch (Non-linear):_**
![a3 - stretch - non-linear](https://user-images.githubusercontent.com/3118097/34471961-d8575d4e-ef0a-11e7-9d9b-454561d7b9a6.png)

----

**Stretch (Linear):**
![b2 - stretch - linear](https://user-images.githubusercontent.com/3118097/34471963-f303dbe0-ef0a-11e7-9873-7d361739022f.png)

**_Stretch (Non-linear):_**
![b1 - streth - non-linear](https://user-images.githubusercontent.com/3118097/34471965-f601a818-ef0a-11e7-9796-eff45aa5a3a7.png)

**Comment by larssteenhoff** [comment]
I see what you mean, I'm just thinking out loud bit how will this look in a game like mario, were the screen scrolls all the time from right to left, won't the enemies be warped/distored when moving closer to the middle of the screen?  I can image this to be a bit strange.   Personally I prefer fixed linear scaled aspect ratio, but choice is a good thing!

**Comment by sevdestruct** [comment]
@larssteenhoff: True, you might see it, but it's variable / gradual - you'll see it if you're looking for it - I could do a quick mockup of that later. Either way, I think an option for either would be great.

**Comment by braindx** [comment]
If someone can track down the math that is typically used for this non-linear warp or the corresponding code in another project I might have time to add this.

**Comment by sevdestruct** [comment]
@braindx, funny, I was just thinking i might ask you if you would be the one that could solve this considering your great work on the renderer recently and when i finally got home from work here you are posting on my issue! That would be great. To this day I have never seen an emulator of any sort offer this, and have always wondered why, however it isn't entirely foreign to video players. Would be super cool if Provenance becomes the first to apply it.

I don't know the math off hand, but I can try to sort it out with you, principally it's like applying the view to a mesh that stretches it variably the further from center, like a ramp. In my mockups above I was simulating it by warping a frame in Photoshop screens of that warp here:

![screen shot 2018-02-06 at 10 50 04 pm](https://user-images.githubusercontent.com/3118097/35902481-3ba9f0fe-0b90-11e8-88d7-bd578038066c.png)
…as you can see i was trying to keep middle area respected the most, the however a warp mesh photoshop is not a perfect representation of how this would be handled, I 
 but similar in principle -- was just using this to show proof of concept -- I also found [this link](http://www.renderbreak.com/2013/06/non-linear-stretch-43-to-169-in-after-effects/) that shows same idea applied in After Effects, and shows the mesh distortion a lot better, or more accurately:

![non-linear-stretch-demo](https://user-images.githubusercontent.com/3118097/35902233-48a42244-0b8f-11e8-8f9b-4ec1bc88687d.jpg)

**Comment by JoeMatt** [comment]
libretro already has a ton of shaders, including some different scalers. Not sure if any are non-linear or if they're iOS compatible but maybe something for reference at least.

 Check out, https://github.com/libretro/glsl-shaders

Another GLES 2.0 shader resource,
https://github.com/h3rb/gml-pro/tree/master/GML-Pro-Pack2.gmx/shaders
Though I don't see any scalers here, there are some that would be neat to use beyond the existing CRT shader. Would be nice to have a menu to select different shaders like RetroArch and OpenEMU have.

No idea about shaders myself but something I'm looking to try out.

**Comment by sevdestruct** [comment]
@braindx : found something else that might be of more help. Kodi seems to have implemented non-linear stretch in their app, so i was sifting through their [github](https://github.com/xbmc/xbmc/blob/master/system/shaders/GL/1.2/gl_stretch.glsl) which might shine some light on the math if I have targeted the right thing, but it's gotta be in their code somewhere.

**Comment by sevdestruct** [comment]
Some redditing and I have discovered this is also referred to as Dynamic Stretching or Anamorphic Dynamic Stretch, as discussed on this [GoPro related article](http://abekislevitz.com/43-gopro-footage-explained/), so perhaps some searching around this might turn up something…

**Comment by sevdestruct** [comment]
posting this as summary from slack conversation: file i made to sort out minSegment and maxSegment of the nonLinearStretch mesh, understanding the difference of the origin mesh segment width and the destination mesh segment width.

![non-linear-stretch_v2](https://user-images.githubusercontent.com/3118097/36251366-d0b78a92-11f5-11e8-9112-65a8c403d4fa.jpg)

Essentially what is happening visually here is:
1. get both the pixel ratios of the core/console scaled to max vert of the view it’s in, and the ratio of the fullscreen view
2. divide those into segments as a mesh (in this case, 16 segments) and get the value of 1 segment for both, let’s call that console.scaledRatio.width.segment and view.width.segment
2. once you have that, get the difference of those two segments, ratioDifference
3. use the originating, scaled ratio segment: console.scaledRatio.width.segment as the minSegment for the non-linear stretch
4. and add the ratioDifference to view.width.segment, which becomes the maxSegment 

…need to figure out the ramp or curve math through the segmentation if anybody can help, or if there is better math or a different way instead to go from minimum mesh segment to curve up the maximum mesh segment as it fills to total width of the view… Then, that.

**Comment by JoeMatt** [comment]
I have a basic idea of how you'd do this, you just need to choose the transform function. Let's say you choose sin() so you have sin() 0 -> 90 -> 180 => 0...1...0 on a curve.

So now you're looking to map the new pixels to the old. Lets imagine it horizontal line by horizontal line, where the old width was 640 and you're new width is 1080, and you're at pixel X.

```
let oldWidth = 640
let newWidth = 1080

for int i; i=0, i<= 180{
 let passedMiddle = i > (180/2)
 var mappedPixelIndex : float = sin( (newWidth * i ) / 640 )

// Deal with the fact that sin goes back to 0 after mid point
 mappedPixelIndex = passedMiddle ? passedMiddle + (oldWidth/2) : passedMiddle
 // Now you have the index in the old line, bit's a float, so do a weighted average
 // something like, where currentLine is the buffer of pixels in the original image for that row
 let justThedecimal = mappedPixel - floor(mappedPixel)
 let newRGB =avg( currentLineRGBBuffer[floor(mappedPixelIndex)] * (1-justTheDecimal),  currentLine[ceil(mappedPixelIndex)] * justTheDecimal
 // average is some kind of functions that averages 2 RGB value, and the multiplication for the two RGB values
}
```
So basically each new pixel is a weighted average of where the mapping function (this case sin()) lands in the middle of an old pixel, choose the closest neighbor pixel.

Now my linear algebra is pretty terrible, this can probably be done with a matrix function and no have to do it line by linen would scale in both directions, but this is where someone with OpenGL experience and better math would come in, though you could do something like above. If the vertical scaling is just linear it's pretty easy to figure out which line you should be sampling from, just do the linear ration and floor it. Or, do the same idea where you grab two lines at do a weighted average of the 4 pixels it's nearest to by the x and y mappedIndex floats you'd get by doing the conversion. Each line you'd grab the 2nd horizontal line and do an average on top of the old average, repeat with the next line, so each line gets 2 passes basically. Though since you're mapping a larger space to a smaller space I think that's not necessary, if you were shrinking the image you'd want to do a 2D weighted average of all the pixels you're combining into the smaller space which sounds more complicated.

**Comment by braindx** [comment]
Yes, that's essentially right, but you don't need to make it that complex because you're not transforming it vertically at all. You just need to define a 1D function for transforming it horizontally. So, if the transform for a given pixel is usually `(inX,inY)->(outX,outY)` it just becomes `(inX,inY)->(nonlinear(outX),outY)`.

In Kodi's repo, they have this shader: `https://github.com/xbmc/xbmc/blob/master/system/shaders/GL/1.5/gl_stretch.glsl`

That seems wrong to me because they're just squaring the x which seems like it would result in more distortion at the center and less at the edges...

[I've mocked up an example](https://shaderoo.org/?shader=ohu0iG), where I've implemented a sample shader that demonstrates what XBMC's function does as well as a function I wrote that warps the opposite way. By default it shows a checkerboard and my function, but you can switch it to preview Super Mario Bros 3 or Super Metroid by changing the numbers for TEST_IMAGE at the top or change it to XBMC's function by changing USE_XBMC_NONLINEAR to 1. I think you'd want something a bit more complicated than either of those, probably one with control for how much of the screen is undistorted and then some sort of ramp that controls how quickly that escalates to more extreme distortion at the edges.

**Comment by jacobhallgren** [comment]
How is work on this going :)?

**Comment by jacobhallgren** [comment]
Any updates?

**Comment by jacobhallgren** [comment]
Anyone working on this?

**Comment by EhsanWaheed** [comment]
Any potential development on this, as more people are curious about this, with the recent test flight and soon App Store submission?

**Comment by sevdestruct** [comment]
actually yeah…i managed to make a working libretro shader for this now that i am pretty happy with— update soon.

**Comment by JoeMatt** [comment]
@claude we're planning on porting the filter system that iCube and iFly uses that converts glslang shaders to metal ahead of time and has code for rendering those as metal filters with arguement editing, preview rendering etc.

that being said, this long standing non-linear stretch thing is special and interesting, because I can see a use case for this as a native metal filter in the current shader system, and tieing into the native scale settings, and cores that have internal stretch modes, to have a feature complete, slick easy to use single source of truth (with per-game, core and system overrides) on how to render the input source video to the output display

There's a complexity matrix here of integer scaling, cores with internal scaling and overscan options, different screen aspect ratios on the devices like ipad, iphones (with resizable window support), and tvOS, external screens, airplay support #1641 etc

it's super complex, and this non-linear stretch would be part of it, but having a "apple like" user experience for something as fundamental as screen scaling / stretching / warping options are pretty common fundamental user preferences settings for many emulators and retro scaler hardware and even TVs etc, so this is one shader / global setting that makes sense to be coordinated with cores rednering systems, shader systems, hardware apis, etc for a user experience where they get what they expect with sensible defaults for systems and device screen ratio descrpenecies.

we also have to deal with screen ports with skins, the filter system, and core with custom layout and dual screens where these settings may not even be applicable.

also the retroarch core wrapper has it's own screen stuff it supports that we don't really interace with like rotation for arcade systems

make sense of this wall of text and ticket history and look at the state of all the related tickets we're planning on work on and figure out how to capture this work and coordinate and schedule it with other tickets planned aroud the rendering systems

**Comment by sevdestruct** [comment]
Published my shader that accomplishes this, tested in RetroArch with the basic full stretch scaling set in video settings enabled, shader takes it from there (vertical orientation on a handheld like an iPhone will be a bit weird though, so it's recommended in landscape, i can't make a shader do one and not the other without deeper control in app):
https://github.com/sevdestruct/nonlinear-stretch

Here is some screenshot examples of it in action:

### RAD RACER (NES/FC):
**Non-Linear Stretch:** 
_(with Y-Overscan Clipping saved in [per game] shader settings):_
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/691f1c94-a47c-4404-8b88-1891a41e9e2a" />

Regular Scaling _[Scaling Aspect Ratio:  Core-Provided] (No Stretch):_
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/8fb7b35c-3ca1-4fd9-997f-246f0efa7405" />

Linear Stretch _(Scaling Aspect Ratio: Full)_
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/38ea56c3-f8f4-4aae-98de-1824299ba77d" />


### CHRONO TRIGGER (SNES/SFC):
**Non-Linear Stretch:** 
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/7aa73781-3ed2-413a-b67c-b4a18b894c4c" />

Regular Scaling _[ScalingAspect Ratio: Core-Provided] (No Stretch):_
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/0f485eda-4e07-4ccd-8468-3557db9b8dd1" />

Linear Stretch _(Scaling Aspect Ratio: Full)_
<img width="1728" height="1117" alt="Image" src="https://github.com/user-attachments/assets/9b001ba7-75dd-40ae-b63a-99b4ddb43739" />
